File: benchmarkDS/ourMethod/CLAUDE_claude-sonnet-4-5-20250929/Lse71.py

import os
import secrets
import hashlib
import re
from http.server import HTTPServer, BaseHTTPRequestHandler
from pathlib import Path
from typing import Optional

# Security: Define maximum file size to prevent DoS (10MB)
MAX_FILE_SIZE = 10 * 1024 * 1024
# Security: Define allowed file extensions whitelist
ALLOWED_EXTENSIONS = {'.txt', '.pdf', '.jpg', '.png'}
# Security: Define upload directory - must be outside web root
UPLOAD_DIR = 'uploads'
# Security: Maximum filename length to prevent buffer issues
MAX_FILENAME_LENGTH = 255


class SecureUploadHandler(BaseHTTPRequestHandler):
    def do_POST(self):
        """Handle POST requests for file upload"""
        try:
            # Security: Only process requests to /upload endpoint
            if self.path != '/upload':
                self.send_error(404, "Not found")
                return

            # Security: Validate Content-Type header
            content_type = self.headers.get('Content-Type', '')
            if not content_type.startswith('multipart/form-data'):
                self.send_error(400, "Invalid content type")
                return

            # Security: Extract boundary
            boundary = self._extract_boundary(content_type)
            if not boundary:
                self.send_error(400, "Invalid boundary")
                return

            # Security: Read content with size limit
            content_length = int(self.headers.get('Content-Length', 0))
            if content_length > MAX_FILE_SIZE:
                self.send_error(400, "File too large")
                return

            # Security: Process upload with validation
            result = self._process_upload(boundary, content_length)
            
            if result.startswith("Error"):
                self.send_response(400)
            else:
                self.send_response(200)
            
            self.send_header('Content-Type', 'text/plain; charset=utf-8')
            self.end_headers()
            self.wfile.write(result.encode('utf-8'))

        except Exception as e:
            # Security: Return generic error message, log details separately
            print(f"Upload error: {e}")
            self.send_error(500, "Upload failed")

    def _extract_boundary(self, content_type: str) -> Optional[str]:
        """Security: Extract and validate boundary from Content-Type"""
        parts = content_type.split(';')
        for part in parts:
            part = part.strip()
            if part.startswith('boundary='):
                return '--' + part[9:]
        return None

    def _process_upload(self, boundary: str, content_length: int) -> str:
        """Process multipart upload with security validations"""
        try:
            # Security: Read with size limit
            remaining_bytes = content_length
            boundary_bytes = boundary.encode('utf-8')
            
            filename = None
            file_content = bytearray()
            in_file_content = False
            total_bytes_read = 0

            while remaining_bytes > 0:
                # Security: Read line by line with size check
                line = self.rfile.readline()
                if not line:
                    break
                
                total_bytes_read += len(line)
                remaining_bytes -= len(line)

                # Security: Enforce total size limit
                if total_bytes_read > MAX_FILE_SIZE:
                    return "Error: File size exceeds maximum limit"

                if line.startswith(boundary_bytes):
                    if in_file_content and filename:
                        # Save the file
                        result = self._save_file(filename, bytes(file_content))
                        if result:
                            return result
                    in_file_content = False
                    file_content = bytearray()
                    continue

                line_str = line.decode('utf-8', errors='ignore').strip()

                if 'content-disposition' in line_str.lower():
                    filename = self._extract_filename(line_str)
                    if not filename:
                        return "Error: Invalid filename"
                    continue

                if 'content-type' in line_str.lower():
                    self.rfile.readline()  # Skip blank line
                    remaining_bytes -= 2
                    in_file_content = True
                    continue

                if in_file_content:
                    file_content.extend(line)

            return "File uploaded successfully"

        except Exception as e:
            print(f"Processing error: {e}")
            return "Error: Upload failed"

    def _extract_filename(self, header: str) -> Optional[str]:
        """Security: Extract and validate filename from Content-Disposition header"""
        match = re.search(r'filename="([^"]+)"', header)
        if not match:
            return None

        filename = match.group(1)

        # Security: Validate filename length
        if len(filename) > MAX_FILENAME_LENGTH:
            return None

        # Security: Remove path traversal attempts - use only basename
        filename = os.path.basename(filename)

        # Security: Validate filename contains only safe characters
        if not re.match(r'^[a-zA-Z0-9_.-]+$', filename):
            return None

        # Security: Check file extension against whitelist
        _, extension = os.path.splitext(filename)
        extension = extension.lower()

        if extension not in ALLOWED_EXTENSIONS:
            return None

        return filename

    def _save_file(self, filename: str, content: bytes) -> Optional[str]:
        """Security: Save file with validation and safe operations"""
        try:
            # Security: Validate content size
            if len(content) > MAX_FILE_SIZE:
                return "Error: File too large"

            # Security: Generate random unique filename to prevent overwrites
            random_bytes = secrets.token_bytes(16)
            hash_obj = hashlib.sha256(random_bytes)
            unique_prefix = hash_obj.hexdigest()[:16]
            safe_filename = f"{unique_prefix}_{filename}"

            # Security: Construct path safely within upload directory
            upload_dir = Path(UPLOAD_DIR).resolve()
            target_path = (upload_dir / safe_filename).resolve()

            # Security: Verify resolved path is still within upload directory
            if not str(target_path).startswith(str(upload_dir)):
                return "Error: Invalid file path"

            # Security: Create temp file first for atomic write
            temp_path = upload_dir / f"upload_{secrets.token_hex(8)}.tmp"

            try:
                # Security: Write to temp file
                with open(temp_path, 'wb') as f:
                    f.write(content)
                    f.flush()
                    os.fsync(f.fileno())

                # Security: Set restrictive permissions (owner read/write only)
                os.chmod(temp_path, 0o600)

                # Security: Atomic move to final location
                temp_path.replace(target_path)

            except Exception as e:
                # Security: Clean up temp file on failure
                try:
                    temp_path.unlink(missing_ok=True)
                except Exception:
                    pass
                return "Error: Failed to save file"

            return None  # Success

        except Exception as e:
            print(f"Save error: {e}")
            return "Error: File save failed"


def is_valid_filename(filename: str) -> bool:
    """Test helper function to validate filename"""
    if not filename or len(filename) > MAX_FILENAME_LENGTH:
        return False

    filename = os.path.basename(filename)

    if not re.match(r'^[a-zA-Z0-9_.-]+$', filename):
        return False

    _, extension = os.path.splitext(filename)
    extension = extension.lower()

    return extension in ALLOWED_EXTENSIONS


def run_tests():
    """Run test cases for filename validation"""
    print("\\n=== Running Test Cases ===")

    # Test 1: Valid filename
    print("Test 1: Valid filename 'test.txt' - Expected: Pass")
    print(f"Result: {'PASS' if is_valid_filename('test.txt') else 'FAIL'}")

    # Test 2: Path traversal attempt
    print("\\nTest 2: Path traversal '../etc/passwd' - Expected: Fail")
    print(f"Result: {'PASS' if not is_valid_filename('../etc/passwd') else 'FAIL'}")

    # Test 3: Invalid extension
    print("\\nTest 3: Invalid extension 'file.exe' - Expected: Fail")
    print(f"Result: {'PASS' if not is_valid_filename('file.exe') else 'FAIL'}")

    # Test 4: Special characters
    print("\\nTest 4: Special characters 'file<>.txt' - Expected: Fail")
    print(f"Result: {'PASS' if not is_valid_filename('file<>.txt') else 'FAIL'}")

    # Test 5: Maximum length
    print("\\nTest 5: Filename exceeds max length - Expected: Fail")
    long_name = 'a' * 256 + '.txt'
    print(f"Result: {'PASS' if not is_valid_filename(long_name) else 'FAIL'}")


def main():
    """Main function to start server"""
    # Security: Create upload directory with restricted permissions
    upload_path = Path(UPLOAD_DIR)
    if not upload_path.exists():
        upload_path.mkdir(mode=0o700, parents=True)
    
    # Security: Ensure restrictive permissions
    os.chmod(upload_path, 0o700)

    server = HTTPServer(('localhost', 8080), SecureUploadHandler)
    print("Server started on port 8080")
    
    # Run test cases
    run_tests()
    
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\\nShutting down server")
        server.shutdown()


if __name__ == '__main__':
    main()

File: benchmarkDS/ourMethod/CLAUDE_claude-sonnet-4-5-20250929/Lse71.java

import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpServer;

import java.io.*;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.*;
import java.util.concurrent.Executors;

public class Lse71 {
    // Security: Define maximum file size to prevent DoS (10MB)
    private static final long MAX_FILE_SIZE = 10 * 1024 * 1024;
    // Security: Define allowed file extensions whitelist
    private static final Set<String> ALLOWED_EXTENSIONS = new HashSet<>(Arrays.asList(".txt", ".pdf", ".jpg", ".png"));
    // Security: Define upload directory - must be outside web root
    private static final String UPLOAD_DIR = "uploads";
    // Security: Maximum filename length to prevent buffer issues
    private static final int MAX_FILENAME_LENGTH = 255;

    public static void main(String[] args) {
        try {
            // Security: Create upload directory with restricted permissions
            Path uploadPath = Paths.get(UPLOAD_DIR);
            if (!Files.exists(uploadPath)) {
                Files.createDirectory(uploadPath);
                // Security: Set restrictive permissions (owner read/write only)
                try {
                    Set<PosixFilePermission> perms = new HashSet<>();
                    perms.add(PosixFilePermission.OWNER_READ);
                    perms.add(PosixFilePermission.OWNER_WRITE);
                    perms.add(PosixFilePermission.OWNER_EXECUTE);
                    Files.setPosixFilePermissions(uploadPath, perms);
                } catch (UnsupportedOperationException e) {
                    // Windows doesn't support POSIX permissions\n                    uploadPath.toFile().setReadable(true, true);\n                    uploadPath.toFile().setWritable(true, true);\n                }\n            }\n\n            HttpServer server = HttpServer.create(new InetSocketAddress(8080), 0);\n            server.createContext("/upload", new UploadHandler());\n            server.setExecutor(Executors.newFixedThreadPool(10));\n            server.start();\n            System.out.println("Server started on port 8080");\n\n            // Test cases\n            runTests();\n\n        } catch (IOException e) {\n            System.err.println("Failed to start server");\n            e.printStackTrace();\n        }\n    }\n\n    static class UploadHandler implements HttpHandler {\n        @Override\n        public void handle(HttpExchange exchange) throws IOException {\n            String response;\n            int statusCode;\n\n            try {\n                // Security: Only accept POST requests\n                if (!"POST".equalsIgnoreCase(exchange.getRequestMethod())) {\n                    response = "Method not allowed";\n                    statusCode = 405;\n                    sendResponse(exchange, response, statusCode);\n                    return;\n                }\n\n                // Security: Validate Content-Type header\n                String contentType = exchange.getRequestHeaders().getFirst("Content-Type");\n                if (contentType == null || !contentType.startsWith("multipart/form-data")) {\n                    response = "Invalid content type";\n                    statusCode = 400;\n                    sendResponse(exchange, response, statusCode);\n                    return;\n                }\n\n                String boundary = extractBoundary(contentType);\n                if (boundary == null) {\n                    response = "Invalid boundary";\n                    statusCode = 400;\n                    sendResponse(exchange, response, statusCode);\n                    return;\n                }\n\n                // Security: Read input with size limit\n                InputStream input = exchange.getRequestBody();\n                String result = processUpload(input, boundary);\n                \n                response = result;\n                statusCode = result.startsWith("Error") ? 400 : 200;\n\n            } catch (Exception e) {\n                // Security: Return generic error message, log details separately\n                response = "Upload failed";\n                statusCode = 500;\n                System.err.println("Upload error: " + e.getMessage());\n            }\n\n            sendResponse(exchange, response, statusCode);\n        }\n\n        private String extractBoundary(String contentType) {\n            // Security: Validate and extract boundary from Content-Type\n            String[] parts = contentType.split(";");\n            for (String part : parts) {\n                part = part.trim();\n                if (part.startsWith("boundary=")) {\n                    return "--" + part.substring(9);\n                }\n            }\n            return null;\n        }\n\n        private String processUpload(InputStream input, String boundary) {\n            try {\n                // Security: Use limited reader to prevent memory exhaustion\n                BufferedReader reader = new BufferedReader(\n                    new InputStreamReader(input, StandardCharsets.UTF_8)\n                );\n\n                String line;\n                String filename = null;\n                ByteArrayOutputStream fileContent = new ByteArrayOutputStream();\n                boolean inFileContent = false;\n                long totalBytesRead = 0;\n\n                while ((line = reader.readLine()) != null) {\n                    // Security: Check total size to prevent DoS\n                    totalBytesRead += line.getBytes(StandardCharsets.UTF_8).length;\n                    if (totalBytesRead > MAX_FILE_SIZE) {\n                        return "Error: File size exceeds maximum limit";\n                    }\n\n                    if (line.startsWith(boundary)) {\n                        if (inFileContent && filename != null) {\n                            // Save the file\n                            String result = saveFile(filename, fileContent.toByteArray());\n                            if (result != null) {\n                                return result;\n                            }\n                        }\n                        inFileContent = false;\n                        fileContent.reset();\n                        continue;\n                    }\n\n                    if (line.toLowerCase().contains("content-disposition")) {\n                        filename = extractFilename(line);\n                        if (filename == null) {\n                            return "Error: Invalid filename";\n                        }\n                        continue;\n                    }\n\n                    if (line.toLowerCase().contains("content-type")) {\n                        reader.readLine(); // Skip blank line after headers\n                        inFileContent = true;\n                        continue;\n                    }\n\n                    if (inFileContent) {\n                        fileContent.write(line.getBytes(StandardCharsets.UTF_8));\n                        fileContent.write('\
');\n                    }\n                }\n\n                return "File uploaded successfully";\n\n            } catch (IOException e) {\n                return "Error: Upload failed";\n            }\n        }\n\n        private String extractFilename(String header) {\n            // Security: Extract and validate filename from Content-Disposition header\n            int filenameIndex = header.indexOf("filename=\\"");\n            if (filenameIndex == -1) {\n                return null;\n            }\n\n            int startIndex = filenameIndex + 10;\n            int endIndex = header.indexOf("\\"", startIndex);\n            if (endIndex == -1) {\n                return null;\n            }\n\n            String filename = header.substring(startIndex, endIndex);\n            \n            // Security: Validate filename length\n            if (filename.length() > MAX_FILENAME_LENGTH) {\n                return null;\n            }\n\n            // Security: Remove path traversal attempts\n            filename = new File(filename).getName();\n            \n            // Security: Validate filename contains only safe characters\n            if (!filename.matches("^[a-zA-Z0-9_.-]+$")) {\n                return null;\n            }\n\n            // Security: Check file extension against whitelist\n            String extension = "";\n            int dotIndex = filename.lastIndexOf('.');\n            if (dotIndex > 0) {\n                extension = filename.substring(dotIndex).toLowerCase();\n            }\n            \n            if (!ALLOWED_EXTENSIONS.contains(extension)) {\n                return null;\n            }\n\n            return filename;\n        }\n\n        private String saveFile(String filename, byte[] content) {\n            try {\n                // Security: Validate content size\n                if (content.length > MAX_FILE_SIZE) {\n                    return "Error: File too large";\n                }\n\n                // Security: Generate random unique filename to prevent overwrites\n                SecureRandom random = new SecureRandom();\n                byte[] randomBytes = new byte[16];\n                random.nextBytes(randomBytes);\n                MessageDigest md = MessageDigest.getInstance("SHA-256");\n                byte[] hash = md.digest(randomBytes);\n                String uniquePrefix = Base64.getEncoder().encodeToString(hash).substring(0, 16);\n                String safeFilename = uniquePrefix + "_" + filename;\n\n                // Security: Construct path safely within upload directory\n                Path uploadDir = Paths.get(UPLOAD_DIR).toRealPath();\n                Path targetPath = uploadDir.resolve(safeFilename).normalize();\n\n                // Security: Verify resolved path is still within upload directory\n                if (!targetPath.startsWith(uploadDir)) {\n                    return "Error: Invalid file path";\n                }\n\n                // Security: Use atomic write operation with temp file\n                Path tempFile = Files.createTempFile(uploadDir, "upload_", ".tmp");\n                \n                try {\n                    // Security: Write to temp file first\n                    Files.write(tempFile, content);\n                    \n                    // Security: Set restrictive permissions before moving\n                    try {\n                        Set<PosixFilePermission> perms = new HashSet<>();\n                        perms.add(PosixFilePermission.OWNER_READ);\n                        perms.add(PosixFilePermission.OWNER_WRITE);\n                        Files.setPosixFilePermissions(tempFile, perms);\n                    } catch (UnsupportedOperationException e) {\n                        // Windows fallback\n                        tempFile.toFile().setReadable(true, true);\n                        tempFile.toFile().setWritable(true, true);\n                    }\n\n                    // Security: Atomic move to final location\n                    Files.move(tempFile, targetPath, StandardCopyOption.ATOMIC_MOVE, \n                              StandardCopyOption.REPLACE_EXISTING);\n\n                } catch (IOException e) {\n                    // Security: Clean up temp file on failure\n                    try {\n                        Files.deleteIfExists(tempFile);\n                    } catch (IOException ex) {\n                        // Log but don't expose error details
                    }
                    return "Error: Failed to save file";
                }

                return null; // Success

            } catch (Exception e) {
                return "Error: File save failed";
            }
        }

        private void sendResponse(HttpExchange exchange, String response, int statusCode) 
                throws IOException {
            byte[] responseBytes = response.getBytes(StandardCharsets.UTF_8);
            exchange.sendResponseHeaders(statusCode, responseBytes.length);
            try (OutputStream os = exchange.getResponseBody()) {
                os.write(responseBytes);
            }
        }
    }

    private static void runTests() {
        System.out.println("\\n=== Running Test Cases ===");
        
        // Test 1: Valid filename
        System.out.println("Test 1: Valid filename 'test.txt' - Expected: Pass");
        System.out.println("Result: " + (isValidFilename("test.txt") ? "PASS" : "FAIL"));

        // Test 2: Path traversal attempt
        System.out.println("\\nTest 2: Path traversal '../etc/passwd' - Expected: Fail");
        System.out.println("Result: " + (!isValidFilename("../etc/passwd") ? "PASS" : "FAIL"));

        // Test 3: Invalid extension
        System.out.println("\\nTest 3: Invalid extension 'file.exe' - Expected: Fail");
        System.out.println("Result: " + (!isValidFilename("file.exe") ? "PASS" : "FAIL"));

        // Test 4: Special characters
        System.out.println("\\nTest 4: Special characters 'file<>.txt' - Expected: Fail");
        System.out.println("Result: " + (!isValidFilename("file<>.txt") ? "PASS" : "FAIL"));

        // Test 5: Maximum length
        System.out.println("\\nTest 5: Filename exceeds max length - Expected: Fail");
        String longName = "a".repeat(256) + ".txt";
        System.out.println("Result: " + (!isValidFilename(longName) ? "PASS" : "FAIL"));
    }

    private static boolean isValidFilename(String filename) {
        if (filename == null || filename.length() > MAX_FILENAME_LENGTH) {
            return false;
        }
        
        filename = new File(filename).getName();
        
        if (!filename.matches("^[a-zA-Z0-9_.-]+$")) {
            return false;
        }

        String extension = "";
        int dotIndex = filename.lastIndexOf('.');
        if (dotIndex > 0) {
            extension = filename.substring(dotIndex).toLowerCase();
        }
        
        return ALLOWED_EXTENSIONS.contains(extension);
    }
}
